src/metabase/models/collection.clj

```diff
@@ -379,31 +426,46 @@
   "Generates an appropriate HoneySQL `:where` clause to filter out descendants of a collection A with a specific property.
   This property is being a descendant of a visible collection other than A. Used for effective children calculations"
   [parent-collection :- CollectionWithLocationAndIDOrRoot, collection-ids :- VisibleCollections]
-  (let [parent-id           (or (:id parent-collection) "")
-        child-literal       (if (collection.root/is-root-collection? parent-collection)
-                              "/"
-                              (format "%%/%s/" (str parent-id)))]
-    (into
-     ;; if the collection-ids are empty, the whole into turns into nil and we have a dangling [:and] clause in query.
-     ;; the (1 = 1) is to prevent this
-     [:and [:= [:inline 1] [:inline 1]]]
-     (if (= collection-ids :all)
-       ;; In the case that visible-collection-ids is all, that means there's no invisible collection ids
-       ;; meaning, the effective children are always the direct children. So check for being a direct child.
-       [[:like :location (h2x/literal child-literal)]]
-       (let [to-disj-ids         (location-path->ids (or (:effective_location parent-collection) "/"))
-             disj-collection-ids (apply disj collection-ids (conj to-disj-ids parent-id))]
-         (for [visible-collection-id disj-collection-ids]
-           [:not-like :location (h2x/literal (format "%%/%s/%%" (str visible-collection-id)))]))))))
-
+  (if (is-trash? parent-collection)
+    ;; the only direct descendants of the Trash are those that were trashed directly. TODO: theoretically if someone
+    ;; Trashes collection A, which I don't have permissions on, and then trashes collection B, which I do - I should
+    ;; see collection B in the Trash. But this is not something we can figure out here, because we need to look at
+    ;; whether a visible ancestor was `trashed_directly`.
+    [:= :trashed_directly true]
+    (let [parent-id           (or (:id parent-collection) "")
+          child-literal       (if (collection.root/is-root-collection? parent-collection)
+                                "/"
+                                (format "%%/%s/" (str parent-id)))
+          collection-ids      (if (= collection-ids :all)
+                                (t2/select-pks-set :model/Collection :archived (:archived parent-collection))
+                                collection-ids)]
+      (into
+       ;; if the collection-ids are empty, the whole into turns into nil and we have a dangling [:and] clause in query.
+       ;; the (1 = 1) is to prevent this
+       [:and [:= [:inline 1] [:inline 1]]]
+       (if (= collection-ids :all)
```

Comment:
This is never true, right? Because we already check if `collection-ids` is `:all` on L439 and then shadow the `collection-ids` var with the results of a t2 call

Reply:
Yep, good call - thanks.